analysis/runtime/src/runtime/global_runtime.rs

```diff
@@ -0,0 +1,75 @@
+use once_cell::sync::OnceCell;
+
+use crate::events::Event;
+
+use super::{scoped_runtime::Runtime, AnyError};
+
+pub struct GlobalRuntime {
+    runtime: OnceCell<Runtime>,
+}
+
+impl GlobalRuntime {
+    /// Create a new [`GlobalRuntime`].
+    ///
+    /// This is not `pub` because it should only be called once below to create [`RUNTIME`],
+    /// as it depends on [`super::FINISHED`], which is also a global.
+    const fn new() -> Self {
+        Self {
+            runtime: OnceCell::new(),
+        }
+    }
+
+    /// Send an [`Event`] to the [`GlobalRuntime`].
+    ///
+    /// If the [`Runtime`] has been initialized, then it sends the [`Event`] to it,
+    /// panicking if there is a [`SendError`](std::sync::mpsc::SendError).
+    ///
+    /// If the [`Runtime`] has not yet been initialized,
+    /// which is done so by [`crate::initialize`],
+    /// which is called at the start of `main`,
+    /// then this means we are executing pre-`main`,
+    /// perhaps in an `.init_array` or `.ctors` section,
+    /// and we cannot use threads yet as the Rust runtime is not fully initialized yet.
+    /// Therefore, we silently drop the [`Event`].
+    /// 
+    /// It also silently drops the [`Event`] if the [`Runtime`]
+    /// has been [`Runtime::finalize`]d/[`GlobalRuntime::finalize`]d.
+    pub fn send_event(&self, event: Event) {
+        match self.runtime.get() {
+            None => {
+                // Silently drop the [`Event`] as the [`Runtime`] isn't ready/initialized yet.
```

Comment:
could we add some debug logs here to get an account of what isn't getting captured, and do you think that's useful?

Reply:
With the old `lighttpd` that still had true pre-`main` static initializers, we skip 1801 `Event`s before `main`, and nothing after `main`.

Reply:
can you collect them in a structure that you only output once main has started?

Reply:
Perhaps what would be safer and still useful would be to simply count the `Event`s and once `main` starts, print how many were skipped if any were.  That way you still get notified it happens and at all, don't drown in tons of `Event`s, and stay safe before `main`.

Alternatively, we could just write the pre-`main` `Event`s on the main thread, perhaps unbuffered, as `write(2)` should always be safe.  Allocation could potentially be unsafe if the allocator uses any `pthread` locks internally.  Furthermore, `std::env::{var,args}{,_os}` I believe are very unsafe pre-`main`, as Rust's runtime that it sets up before `main` explicitly initializes these.

Overall, I think it's best to do the bare minimum before `main` and it should be up to the user to move their static initializers into `main` if they want them analyzed.  I think the most we should do is notify the user that this has happened so they can fix it.

Reply:
> can you collect them in a structure that you only output once main has started?

Oh, sorry, I didn't see this comment until now.

That's not the best option either for a few reasons:
* We'd have to allocate, which likely is but may not be safe.
* We'd need to collect the skipped `Event`s in a `static`, which we'd have to protect with a `Mutex` to make it thread-safe, but we can't use `Mutex`es before `main` on all OSes and Rust versions as that uses `libpthread`.
* Since this code is pre-`main`, we may be able to use `unsafe` and a `static mut` as there should be no other threads.  That still feels pretty dangerous, though.

Reply:
I'm fine with the collected count and debug print. I'm not sure how useful 1800 lines of debug output really would be, and it's not that hard to go remove constructors if someone really wants to capture all these events (like I did for lighttpd).

Reply:
Yeah, that's what I was thinking.  We really just need to notify the user that this is happening so they can fix it and do things correctly.  I'll add this commit then.

Reply:
Fixed in https://github.com/immunant/c2rust/pull/490/commits/1b72fd4ee2c716ab8f3ab9634bca778df96fb4ca.